### DISCUSSION TITLE
Best/any supported way to set up Homebrew for a multi-account but single-user system?

### DISCUSSION THREAD

**Discussion by GottemHams** [Homebrew/discussions#3253] [Getting started] [answered]
I got a new laptop from my new job and I had to use Homebrew to set up a dev environment. I'm also already using Homebrew on my current 100% private Mac for things such as `virt-manager`. Since I'm allowed to use the new laptop for personal stuff too, I figured I'd create 3 separate accounts on it: one for work, one personal and a dedicated one for just Homebrew (to avoid many duplicate packages). Then I just used some `/etc/sudoers` magic to allow executing of `/opt/homebrew/bin/brew` as `mybrew` without a password for the other 2 accounts. However, this results in problems with services because they require an actual GUI login session to even start (`Error: Failure while executing; /bin/launchctl bootstrap gui/503 /Users/mybrew/Library/LaunchAgents/homebrew.mxcl.httpd.plist` etc).

I don't really wanna have to "physically" sign in to the `mybrew` account every time I reboot. I think the above mentioned error is due to it being a Launch**Agent** and those only act when the account is actually logged in (through CoreServices I think?). They may even require a login window because they have access to the window server? I haven't actually tried putting the Apache plist in the **system** `/Library/LaunchAgents` (or even Launch**Daemons**), but those aren't writable as non-root anyways so Homebrew won't be able to adhere to its "no sudo" philosophy. While it might work for now, something will most definitely break when running package upgrades.

I'm also aware of the workaround where you create a new group and add (in my case) all 3 users to it. However, that will still result in permissions issues down the line. Like if I have `mybrew` install a new package, the default permissions for directories will be 644 with an ownership of `mybrew:staff`. Meaning the other 2 accounts won't be able to upgrade/remove a package later, but I actually want to be able to do just that. Sure, I could have a script continuously `chmod`/`chown` things, but that doesn't really feel "clean". Also, I think certain database servers might refuse to run when its data directory has too wide permissions (6**6**4). ACL inheritance on macOS is a bit funky as far as I know, as in it doesn't usually inherit anything and especially when done directly through shell commands instead of Finder/CoreServices.

I'm still the only one who will be using the laptop (hence my term "multi-account but single-user"), so certain privacy problems won't really be an issue for me. Is there any **supported** way to allow both accounts to install packages to a global Homebrew instance? If not, what is the **most reliable** method to work around the issue? Like, something that doesn't require you to fix things every time you run `brew upgrade`, even if that fix would be scriptable. If it matters anything: it's an M1 Mac running Monterey.

**Comment by SMillerDev** [comment]
It seems like the main issue is `brew services` doesn't play nice with other accounts. So why not just try `brew services` from the respective private/work users?

**Reply by GottemHams** [reply]
I think it might still cause permission problems though, e.g. when Apache rotates a log: the account that last started it will own the new log file and the others can't write to it anymore. Even when scripting the permissions change there'll always be a bit of time between rotation and making it writable again. I guess that's my main "problem" with that and why it doesn't really feel "clean", since it wouldn't apply just to log files but *anything* under `/opt/homebrew`.

I looked at `launchctl` some more and there might be a way to do it natively, but it would require a change on the Homebrew side (or at least the formulae). `launchctl` help output includes this:
```
user/<uid>/[service-name]
Targets the user domain or service within. A process running as the target user
may make modifications. Root may modify any user's domain. User domains do not
exist on iOS.

gui/<uid>/[service-name]
Targets the GUI domain or service within. Each GUI domain is associated with a
user domain, and a process running as the owner of that user domain may make
modifications. Root may modify any GUI domain. GUI domains do not exist on iOS.
```

As well as:
```
	bootstrap       Bootstraps a domain or a service into a domain.
	kickstart       Forces an existing service to start.
	load            Bootstraps a service or directory of services.
```

Furthermore, the `man` page for it actually says `load` is a *legacy* subcommand (didn't even know that until now and I still use it myself everywhere lel). The replacement is `launchctl bootstrap gui/503 /Users/mybrew/Library/LaunchAgents/homebrew.mxcl.httpd.plist`, which is exactly what `brew services` chokes on. Now what if Homebrew were to do this `bootstrap` command with the `user/503` domain instead of just `launchctl load` (which always seems to end up doing `bootstrap gui/503`)? *Services/daemons* that don't have a GUI or otherwise require access to the login window might run just fine like that. Doing a quick `grep` it looks like every formula has to specify its `launchctl` command itself anyways? Then everything for which it's known they (might) require a GUI could simply keep using `load` or switch to `bootstrap gui/<UID>`. Of course that means using GUI daemons across multiple accounts still won't work, but how many of those interact with the login window (I know none for macOS)? Normally there's a helper program running in the GUI that interacts with a daemon, which may be running under a different account. Regular programs like `virt-manager` aren't run through `launchctl` or even `brew services` for that matter, so they will always run with the current GUI account.

The `kickstart` command may or may not be necessary so I only added it for completeness.

I haven't actually checked yet if this theory works as described, but might Homebrew be receptive of such a change? It already says `load` is legacy on my current *Mojave* and since Homebrew only supports the most recent major versions of macOS anyways, there should be no problem with compatibility.

**Comment by GottemHams** [comment]
Alright, that seems to work if I make one more change; the `.plist` should also contain this:
```
<key>LimitLoadToSessionType</key>
<string>Background</string>
```

So these are the full steps when using a magic `sudo` command to execute stuff as `mybrew`:
1. `brew services start httpd` so it creates the LaunchAgent
2. `vim /Users/mybrew/Library/LaunchAgents/homebrew.mxcl.httpd.plist` and add `LimitLoadToSessionType`
3. `launchctl bootstrap user/503 /Users/mybrew/Library/LaunchAgents/homebrew.mxcl.httpd.plist`

Which makes Apache start up nicely:
```
$ ps aux | grep [h]ttpd
mybrew           8142   0.0  0.0 408509536   1648   ??  S 12:16PM   0:00.00 /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND
mybrew           8141   0.0  0.0 408509536   1712   ??  S 12:16PM   0:00.00 /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND
mybrew           8140   0.0  0.0 408509536   1728   ??  S 12:16PM   0:00.00 /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND
mybrew           8139   0.0  0.0 408509536   1648   ??  S 12:16PM   0:00.00 /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND
mybrew           8138   0.0  0.0 408509536   1664   ??  S 12:16PM   0:00.00 /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND
mybrew           8134   0.0  0.0 408771984   4992   ??  S 12:15PM   0:00.06 /opt/homebrew/opt/httpd/bin/httpd -D FOREGROUND 
```

I don't have any login sessions open for `mybrew` and I even killed all processes owned by that account *prior* to `launchctl bootstrap`, just to be sure. After the command has long since finished I can still go to my browser and open any page hosted by Apache no problem.

Note: `launchctl load` (or its replacement `bootstrap gui/<UID>`) still no werkies, even with the addition of `LimitLoadToSessionType`. The `gui/*` domain actually doesn't even exist until you're graphically logged on. The `user/*` domain already gets created when SSHing/`su(do)`ing into it and I'm 99% sure that due to `LimitLoadToSessionType` it's allowed to continue running even when the user logs out. Or maybe it just keeps the domain around until you reboot, I haven't actually checked that far.

**Reply by GottemHams** [reply]
@SMillerDev Should I maybe make this an official feature request by opening an issue for it, or are the chances of it getting implemented rather low?

**Reply by SMillerDev** [reply]
Feature requests should really only be filed as a pull request, but if you make one of those I'd be happy to help you and review it.

**Reply by GottemHams** [reply]
Sure, I could take a look sometime.

**Comment by GottemHams** [comment, marked as answer]
Finally took some time to move everything over to the new Mac, so [here's the PR](https://github.com/Homebrew/homebrew-services/pull/485). =]

**Comment by RandomDSdevel** [comment]
> I'm also aware of the workaround where you create a new group and add (in my case) all 3 users to it. However, that will still result in permissions issues down the line. Like if I have `mybrew` install a new package, the default permissions for directories will be 644 with an ownership of `mybrew:staff`. Meaning the other 2 accounts won't be able to upgrade/remove a package later, but I actually want to be able to do just that. Sure, I could have a script continuously `chmod`/`chown` things, but that doesn't really feel "clean". Also, I think certain database servers might refuse to run when its data directory has too wide permissions (664). ACL inheritance on macOS is a bit funky as far as I know, as in it doesn't usually inherit anything and especially when done directly through shell commands instead of Finder/CoreServices.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I've [heard](https://forums.macrumors.com/threads/apple-discontinues-macos-server.2342487/post-31042286) that [TinkerTool](http://www.bresink.com/osx/TinkerTool.html) has settings for automatically re-propagating folders' permissions and ACLs to their contents, likely recursively, if that helps any.
